File: 2020/four/four.py
#!/user/bin/env python3

from sys import argv
import re

def parse(filename):
  db = []
  with open(filename, 'r') as infile:
    for passport in re.split('\n\n',infile.read().strip()):
      entry = {}
      for record in passport.replace('\n',' ').split(' '):
        key = record.split(':')[0]
        value = record.split(':')[1]
        entry[key] = value
      db.append(entry)
  return (db) 

def check(entry):
  fields = ['byr','iyr','eyr','hgt','hcl','ecl','pid']
  if len(entry) < 7:
    return 1 # not enough total fields
  for field in fields:
    if entry.get(field) == None:
      return 2 # missing required fields
  if entry.get('cid') == None:
    return 3 # valid but northpole
  return 0 # normal valid

if __name__ == '__main__':
  db = parse(argv[1])
  for entry in db:
    ret = check(entry)
    if ret == 0:
      print ('Valid')
    elif ret == 3:
      print ('Valid - North Pole')
    elif ret == 1:
      print ('Invalid - not enough fields')
    elif ret == 2:
      print ('Invalid - missing required fields')

File: 2020/four/four.rs
use std::env;
use std::fs::File;
use std::io::{self, BufRead};
use std::path::Path;
use std::collections::HashMap;

fn read_lines<P>(filename: P) -> io::Result<io::Lines<io::BufReader<File>>>
where P: AsRef<Path>, {
  let file = File::open(filename)?;
  Ok(io::BufReader::new(file).lines())
}

fn parse(filename: &String) -> Vec<HashMap<String,String>> {
  let mut db = Vec::new();
  db.push(HashMap::new());
  let mut entry = db.len()-1;
  if let Ok(lines) = read_lines(filename) {
    for line in lines {
      if let Ok(result) = line {
        for record in result.split(' ') {
          for pair in record.split(':').collect::<Vec<&str>>().chunks(2) {
            if pair.len() == 2 {
              db[entry].insert(pair[0].to_string(),pair[1].to_string());
            } else {
              db.push(HashMap::new());
              entry = db.len()-1;
            }
          }
        }
      }
    }
  }
  return db;
}

fn check(record: &HashMap<String,String>) -> Result<&str,&str>
{
  let fields = ["byr","iyr","eyr","hgt","hcl","ecl","pid"];
  if record.len() < 7 {
    return Err("Not enough total fields");
  }
  for field in &fields {
    //if let None(e) = record.get(&field.to_string()) {
    if record.get(&field.to_string()) == None {
      return Err("Missing required fields");
    }
  }
  if record.get("cid") == None {
    return Ok("Valid - North Pole");
  }
  return Ok("Valid");
}

fn main() {
  let args: Vec<String> = env::args().collect();
  let db = parse(&args[1]);
  for record in db {
    println!("{:?}",check(&record));
  }
}
